libs/providers/multi-provider/src/lib/multi-provider.ts

```diff
@@ -0,0 +1,315 @@
+      .map(([_, resolution]) => resolution)
+      .filter((r): r is ProviderResolutionResult<T> => Boolean(r));
+
+    const finalResult = this.evaluationStrategy.determineFinalResult({ flagKey, flagType }, context, resolutions);
+
+    if (finalResult.errors?.length) {
+      throw constructAggregateError(finalResult.errors);
+    }
+
+    if (!finalResult.details) {
+      throw new GeneralError('No result was returned from any provider');
+    }
+
+    return finalResult.details;
+  }
+
+  private async evaluateProviderEntry<T extends boolean | string | number | JsonValue>(
+    flagKey: string,
+    flagType: FlagValueType,
+    defaultValue: T,
+    providerEntry: RegisteredProvider,
+    hookContext: HookContext,
+    hookHints: HookHints,
+    context: EvaluationContext,
+  ): Promise<[shouldEvaluateNext: boolean, ProviderResolutionResult<T> | null]> {
+    let evaluationResult: ResolutionDetails<T> | undefined = undefined;
+    const provider = providerEntry.provider;
+    const strategyContext = {
+      flagKey,
+      flagType,
+      provider,
+      providerName: providerEntry.name,
+      providerStatus: this.statusTracker.providerStatus(providerEntry.name),
+    };
+
+    if (!this.evaluationStrategy.shouldEvaluateThisProvider(strategyContext, context)) {
+      return [true, null];
+    }
+
+    let resolution: ProviderResolutionResult<T>;
+
+    try {
+      evaluationResult = await this.evaluateProviderAndHooks(flagKey, defaultValue, provider, hookContext, hookHints);
+      resolution = {
+        details: evaluationResult,
+        provider: provider,
+        providerName: providerEntry.name,
+      };
+    } catch (error: unknown) {
+      resolution = {
+        thrownError: error,
+        provider: provider,
+        providerName: providerEntry.name,
+      };
+    }
+
+    return [
+      this.evaluationStrategy.runMode === 'sequential'
+        ? this.evaluationStrategy.shouldEvaluateNextProvider(strategyContext, context, resolution)
+        : true,
+      resolution,
+    ];
+  }
+
+  private async evaluateProviderAndHooks<T extends boolean | string | number | JsonValue>(
+    flagKey: string,
+    defaultValue: T,
+    provider: Provider,
+    hookContext: HookContext,
+    hookHints: HookHints,
+  ) {
+    let providerContext: EvaluationContext | undefined = undefined;
+    let evaluationResult: ResolutionDetails<T>;
+
+    try {
+      providerContext = await this.hookExecutor.beforeHooks(provider.hooks, hookContext, hookHints);
+
+      evaluationResult = (await this.callProviderResolve(
+        provider,
+        flagKey,
+        defaultValue,
+        providerContext,
+      )) as ResolutionDetails<T>;
+
+      const afterHookEvalDetails = {
+        ...evaluationResult,
+        flagMetadata: Object.freeze(evaluationResult.flagMetadata ?? {}),
+        flagKey,
+      };
+
+      await this.hookExecutor.afterHooks(
+        provider.hooks,
+        { ...hookContext, context: providerContext },
```

Comment:
This makes it so provider hooks can't use WeakMaps. Would it be possible to use the same object reference for the hook context so that functionality could persist?

Reply:
Here's the code I used to test this.

```ts
import { OpenFeature, InMemoryProvider, Hook, BeforeHookContext, EvaluationContext, EvaluationDetails, FlagValue, HookContext } from "@openfeature/server-sdk";
import { MultiProvider, FirstMatchStrategy } from "@openfeature/multi-provider";

class WeakMapHook implements Hook {
  private map = new WeakMap<WeakKey, boolean>();

  before(hookContext: BeforeHookContext, hookHints?: Readonly<Record<string, unknown>> | undefined): void | EvaluationContext | Promise<void | EvaluationContext> {
      this.map.set(hookContext, true);
  }

  after(hookContext: Readonly<HookContext<FlagValue>>, evaluationDetails: EvaluationDetails<FlagValue>, hookHints?: Readonly<Record<string, unknown>> | undefined): void | Promise<void> {
    if (this.map.has(hookContext)) {
      console.log("WeakMapHook executed");
    } else {
      console.log("WeakMapHook not executed");
    }
  }
}

class CustomInMemoryProvider extends InMemoryProvider {
  hooks = [
    new WeakMapHook(),
  ]

  constructor(flagConfig: ConstructorParameters<typeof InMemoryProvider>[0]) {
    super(flagConfig);
  }
}

await OpenFeature.setProviderAndWait(
  new MultiProvider([
    {
      provider: new InMemoryProvider({
        test1: {
          variants: {
            on: true,
            off: false,
          },
          defaultVariant: "on",
          disabled: false,
        },
      }),
      name: "provider1",
    },
    {
      provider: new InMemoryProvider({
        test2: {
          variants: {
            on: true,
            off: false,
          },
          defaultVariant: "on",
          disabled: false,
        },
      }),
      name: "provider2",
    },
    {
      provider: new CustomInMemoryProvider({
        test3: {
          variants: {
            on: true,
            off: false,
          },
          defaultVariant: "on",
          disabled: false,
        },
      }),
      name: "provider3",
    }
  ], new FirstMatchStrategy())
);

const client = OpenFeature.getClient();

console.log(await client.getBooleanDetails("test", false));
console.log(await client.getBooleanDetails("test1", false));
console.log(await client.getBooleanDetails("test2", false));
console.log(await client.getBooleanDetails("test3", false));
```

Reply:
This is an important point, I think we need to use the same object instance.

Reply:
updated to keep the reference consistent between hook calls of an individual provider. In order to do that I had to make a copy of the hook context + eval context for each provider's evaluation, so the object is consistent for an individual provider but is not consistent _across_ providers.